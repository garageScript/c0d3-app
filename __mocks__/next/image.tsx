import React from 'react'
import type { ImageProps } from 'next/image'
import _ from 'lodash'

const Image = (props: ImageProps) => {
  /*not all Image properties are allowed in <img/> tag
  ObjectFit causes warnings, but otherwise lessons/error images won't keep their dimensions 
  see https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit*/
  const copy = _.omit(props, 'objectFit')
  return <img {...copy} />
}

export default Image
