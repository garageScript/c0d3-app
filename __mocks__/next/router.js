const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
