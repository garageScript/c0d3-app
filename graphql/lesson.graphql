query Lesson {
  lessons {
    id
    title
  }
}