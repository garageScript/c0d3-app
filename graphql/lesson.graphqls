type Lesson {
  id: String
  title: String
}

type Query {
  lessons: Lesson!
}