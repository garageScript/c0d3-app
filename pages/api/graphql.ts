import { ApolloServer } from 'apollo-server-micro'
import session from 'express-session'
import nextConnect from 'next-connect'
import db from '../../helpers/dbload'
import typeDefs from '../../graphql/typeDefs'
import resolvers from '../../graphql/resolvers'
import connectSequelize from 'connect-session-sequelize'

// Sequelize Store for express-sessions
const { sequelize } = db
const SequelizeStore = connectSequelize(session.Store)

const ONE_WEEK = 1000 * 60 * 60 * 24 * 7

const handler: any = nextConnect() // For session middleware. TODO: Need to define types for Session
const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req }) => ({ req }) // This lets GraphQL have access to sessions
})

const graphQLHandler = apolloServer.createHandler({ path: '/api/graphql' })

handler
  .use(
    session({
      secret: process.env.SESSION_SECRET || '',
      store: new SequelizeStore({
        db: sequelize
      }),
      resave: false,
      saveUninitialized: false,
      cookie: {
        domain: process.env.DB_HOST,
        maxAge: ONE_WEEK
      }
    })
  )
  .get('/api/graphql', graphQLHandler)
  .post('/api/graphql', graphQLHandler)

export const config = {
  api: {
    bodyParser: false // This is required for GraphQL to work properly  https://blog.logrocket.com/building-a-graphql-server-in-next-js/
  }
}

export default handler
