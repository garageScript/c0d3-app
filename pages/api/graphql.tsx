import { ApolloServer, gql } from 'apollo-server-micro'

const typeDefs = gql`
  type Query {
    username: String
    hello: String
  }
`

const resolvers = {
  Query: {
    username(_parent: any, _args: any, ctx: any) {
      return ctx.user
    },
    hello() {
      return 'hello'
    }
  }
}

const context = ({ req }: { req: any }) => req.cookies

const apolloServer = new ApolloServer({ typeDefs, resolvers, context })

export const config = {
  api: {
    bodyParser: false
  }
}

export default apolloServer.createHandler({ path: '/api/graphql' })
