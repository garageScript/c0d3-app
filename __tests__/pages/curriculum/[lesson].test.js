import React from 'react'
import { render, waitFor, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../../graphql/queries/getApp'
import Lesson from '../../../pages/curriculum/[lesson]'
import dummyLessonData from '../../../__dummy__/lessonData'
import dummySessionData from '../../../__dummy__/sessionData'
import dummyAlertData from '../../../__dummy__/alertData'
import { useRouter } from 'next/router'
import { SubmissionStatus } from '../../../graphql'

const session = {
  ...dummySessionData,
  submissions: [
    {
      id: '1',
      status: SubmissionStatus.Passed,
      mrUrl: '',
      diff: '',
      viewCount: 0,
      comment: '',
      order: 0,
      challengeId: '146',
      lessonId: '2',
      reviewer: {
        id: '1',
        username: 'fake reviewer'
      },
      createdAt: '123',
      updatedAt: '123'
    },
    {
      id: '1',
      status: SubmissionStatus.Passed,
      mrUrl: '',
      diff: '',
      viewCount: 0,
      comment: '',
      order: 0,
      challengeId: '145',
      lessonId: '2',
      reviewer: {
        id: '1',
        username: 'fake reviewer'
      },
      createdAt: '123',
      updatedAt: '123'
    }
  ],
  lessonStatus: [
    {
      lessonId: '5',
      isPassed: true,
      isTeaching: true,
      isEnrolled: false,
      starGiven: null
    },
    {
      lessonId: '2',
      isPassed: false,
      isTeaching: false,
      isEnrolled: true,
      starGiven: null
    }
  ]
}
describe('Lesson Page', () => {
  const { query } = useRouter()
  query['lesson'] = '2'
  test('Should render correctly with valid lesson route', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session,
            lessons: dummyLessonData,
            alerts: dummyAlertData
          }
        }
      }
    ]

    const { container, getByRole } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Lesson />
      </MockedProvider>
    )

    await waitFor(() =>
      getByRole('heading', { name: /Variables & Functions/i })
    )

    await waitFor(() => expect(container).toMatchSnapshot())
  })

  test('Should render correctly with invalid lesson route', async () => {
    query['lesson'] = '100'
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session,
            lessons: dummyLessonData,
            alerts: dummyAlertData
          }
        }
      }
    ]
    const { container, getByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Lesson />
      </MockedProvider>
    )

    await waitFor(() => getByText(/Page not found/i))

    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test("Should correctly render challenges page for students who hadn't passed previous lessons", async () => {
    query['lesson'] = '25'
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session,
            lessons: dummyLessonData,
            alerts: dummyAlertData
          }
        }
      }
    ]

    const { container, getByRole } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Lesson />
      </MockedProvider>
    )

    await waitFor(() => getByRole('heading', { name: /Trees/i }))

    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should return Internal server Error if alerts or lessons are missing', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session,
            lessons: null,
            alerts: dummyAlertData
          }
        }
      }
    ]

    const { container, findByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Lesson />
      </MockedProvider>
    )

    const element = await findByText(/Internal server error/i)
    expect(element).toBeTruthy()

    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should render with nulled submissions', async () => {
    query['lesson'] = '2'
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: { ...session, submissions: null },
            lessons: dummyLessonData,
            alerts: dummyAlertData
          }
        }
      }
    ]

    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Lesson />
      </MockedProvider>
    )

    await waitFor(() =>
      screen.getByRole('heading', { name: /Variables & Functions/i })
    )
  })
})
