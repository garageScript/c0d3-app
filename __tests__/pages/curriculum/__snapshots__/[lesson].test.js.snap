// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson Page Should correctly render challenges page for students who hadn't passed previous lessons 1`] = `
<div>
  <div>
    <nav
      class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand text-primary font-weight-bold"
          href="/"
        >
          C0D3
        </a>
        <div
          id="navbarNav"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <div
              class="navbar-nav collapse navbar-collapse"
            >
              <a
                class="nav-item nav-link"
                href="/curriculum"
              >
                Curriculum
              </a>
              <a
                class="nav-item nav-link"
                href="https://github.com/garageScript/c0d3-app"
                rel="noopener noreferrer"
                target="_blank"
              >
                Repo
              </a>
              <a
                class="nav-item nav-link"
                href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
                rel="noopener noreferrer"
                target="_blank"
              >
                Journey
              </a>
              <a
                class="nav-item nav-link"
                href="https://chat.c0d3.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                Help
              </a>
              <a
                class="nav-item nav-link"
                href="/contributors"
              >
                Contributors
              </a>
              <div
                class="dropdown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-none"
                  type="button"
                >
                  Admin
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-flex"
        >
          <a
            class="btn btn-secondary border overflow-hidden p-2 text-truncate"
            href="/profile/fakeusername"
          >
             fakeusername
          </a>
          <button
            class="btn border ml-2"
            style="color: rgb(41, 41, 41);"
          >
            Logout
          </button>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="row mt-4"
      >
        <div
          class="challenges-container"
        >
          <div
            class="card shadow-sm mt-3 col-12 px-0 pt-3 border-0"
          >
            <div
              class="card-body p-0"
            >
              <div
                class="d-flex mb-3 px-3"
              >
                <img
                  alt="lesson-cover"
                  class="lessonTitleCard__lesson-cover mr-3"
                  src="/assets/curriculum/js-8-cover.svg"
                />
                <div>
                  <p
                    class="m-0"
                  >
                    <a
                      class=""
                      href="/curriculum"
                    >
                      Go Back
                    </a>
                  </p>
                  <h1
                    class="lessonTitleCard__lesson-title"
                  >
                    Trees
                  </h1>
                </div>
              </div>
              <div
                class="card-footer bg-white p-0"
              >
                <a
                  class="btn border-right rounded-0 px-4 py-3"
                  href="/curriculum/25"
                >
                  CHALLENGES
                </a>
              </div>
            </div>
          </div>
          <div
            class="alerts-container "
          >
            <div
              class="alert d-flex justify-content-between mt-3 bg-primary text-white"
              role="alert"
            >
              <div>
                <img
                  class="mr-3 alert-icon"
                  src="/assets/curriculum/icons/icon-tip.svg"
                />
                Set up your computer to submit challenges. 
                <a
                  class="text-white"
                  href="https://www.notion.so/JS-0-Foundations-a43ca620e54945b2b620bcda5f3cf672#b45ed85a95e24c9d9fb784afb7a46bcc"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  View Instructions
                </a>
              </div>
              <button
                class="alert-dismiss"
                data-testid="dismiss-info"
                role="dismiss"
              >
                <img
                  src="/assets/curriculum/icons/dismiss-info.svg"
                />
              </button>
            </div>
            <div
              class="alert d-flex justify-content-between mt-3 alert-danger"
              role="alert"
            >
              <div>
                <img
                  class="mr-3 alert-icon"
                  src="/assets/curriculum/icons/exclamation.svg"
                />
                Please upgrade your CLI client by running npm update c0d3. 
                
              </div>
              <button
                class="alert-dismiss"
                data-testid="dismiss-urgent"
                role="dismiss"
              >
                <img
                  src="/assets/curriculum/icons/dismiss-urgent.svg"
                />
              </button>
            </div>
          </div>
          <div
            class="row challenge-display mt-3"
          >
            <div
              class="col-4"
            >
              <div
                class="card mb-2 challenge-title-card challenge-title-card--active shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    1. Given a binary tree, find the largest value
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    2. Given a binary tree, find the maximum depth
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    3. Given a binary tree, return the maximum distance between any 2 nodes
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    4. Biggest subtree value
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    5. Given a binary tree and a node, find the distance to that node.
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    6. Max Fun
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    7. Edit distance
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    8. Maximum Loot
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    9. Longest Increasing Non-Consecutive Sequence
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    10. Minimum Coins for Change
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    11. Maximum Weight to Carry
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    12. Subarray Sum Equals K
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col-8"
            >
              <div
                class="card shadow-sm border-0"
              >
                <div
                  class="card-body challenge-question"
                >
                  <h1
                    class="card-title"
                    data-testid="challenge-question-title"
                  >
                    Given a binary tree, find the largest value
                  </h1>
                  <div
                    class="bg-light p-3 mt-3"
                    data-testid="challenge-question-description"
                  >
                    <div>
                      <p>
                        Example:
                      </p>
                      <p>
                        <img
                          alt="tree"
                          src="https://c0d3.com/apis/draw?tree={%22v%22:30,%22children%22:[{%22v%22:50},{%22v%22:40}]}"
                        />
                      </p>
                      <pre>
                        <code>
                          const a = {val: 30, right: {val: 40}, left: {val: 50}};
solution(a); // should return 50
                        </code>
                      </pre>
                    </div>
                  </div>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="mt-5 font-weight-light text-center pb-5 text-muted"
    >
      © Copyright 2020 GarageScript
      <br />
      All Rights Reserved
    </footer>
  </div>
</div>
`;

exports[`Lesson Page Should render correctly with invalid lesson route 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
              rel="noopener noreferrer"
              target="_blank"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
            <a
              class="nav-item nav-link"
              href="/contributors"
            >
              Contributors
            </a>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-none"
                type="button"
              >
                Admin
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex"
      >
        <a
          class="btn btn-secondary border overflow-hidden p-2 text-truncate"
          href="/profile/fakeusername"
        >
           fakeusername
        </a>
        <button
          class="btn border ml-2"
          style="color: rgb(41, 41, 41);"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="d-flex col-sm-3 align-items-center justify-content-center"
        >
          <div
            class="text-center mt-3"
          >
            <div
              style="color: rgb(41, 41, 41); font-size: 1.5rem;"
            >
              <strong>
                Page not found
              </strong>
            </div>
            <div
              style="color: rgb(41, 41, 41); font-size: 1rem;"
            >
              Lesson not found
            </div>
            <a
              class="btn btn-primary py-3 px-5 mt-3"
              href="/"
            >
              <h3
                class="font-weight-bold"
              >
                Back
              </h3>
            </a>
          </div>
        </div>
        <div
          class="mb-3 col-sm-9"
        >
          <div
            style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="display: block; box-sizing: border-box; padding-top: 100%;"
            />
            <img
              decoding="async"
              sizes="100vw"
              src="/_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=3840&q=75"
              srcset="/_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=640&q=75 640w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=750&q=75 750w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=828&q=75 828w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=1080&q=75 1080w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=1200&q=75 1200w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=1920&q=75 1920w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=2048&q=75 2048w, /_next/image?url=%2Fassets%2Ferrors%2F404.svg&w=3840&q=75 3840w"
              style="visibility: inherit; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Lesson Page Should render correctly with valid lesson route 1`] = `
<div>
  <div>
    <nav
      class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand text-primary font-weight-bold"
          href="/"
        >
          C0D3
        </a>
        <div
          id="navbarNav"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <div
              class="navbar-nav collapse navbar-collapse"
            >
              <a
                class="nav-item nav-link"
                href="/curriculum"
              >
                Curriculum
              </a>
              <a
                class="nav-item nav-link"
                href="https://github.com/garageScript/c0d3-app"
                rel="noopener noreferrer"
                target="_blank"
              >
                Repo
              </a>
              <a
                class="nav-item nav-link"
                href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
                rel="noopener noreferrer"
                target="_blank"
              >
                Journey
              </a>
              <a
                class="nav-item nav-link"
                href="https://chat.c0d3.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                Help
              </a>
              <a
                class="nav-item nav-link"
                href="/contributors"
              >
                Contributors
              </a>
              <div
                class="dropdown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-none"
                  type="button"
                >
                  Admin
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-flex"
        >
          <a
            class="btn btn-secondary border overflow-hidden p-2 text-truncate"
            href="/profile/fakeusername"
          >
             fakeusername
          </a>
          <button
            class="btn border ml-2"
            style="color: rgb(41, 41, 41);"
          >
            Logout
          </button>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="row mt-4"
      >
        <div
          class="challenges-container"
        >
          <div
            class="card shadow-sm mt-3 col-12 px-0 pt-3 border-0"
          >
            <div
              class="card-body p-0"
            >
              <div
                class="d-flex mb-3 px-3"
              >
                <img
                  alt="lesson-cover"
                  class="lessonTitleCard__lesson-cover mr-3"
                  src="/assets/curriculum/js-1-cover.svg"
                />
                <div>
                  <p
                    class="m-0"
                  >
                    <a
                      class=""
                      href="/curriculum"
                    >
                      Go Back
                    </a>
                  </p>
                  <h1
                    class="lessonTitleCard__lesson-title"
                  >
                    Variables & Functions
                  </h1>
                </div>
              </div>
              <div
                class="card-footer bg-white p-0"
              >
                <a
                  class="btn border-right rounded-0 px-4 py-3"
                  href="https://www.notion.so/garagescript/JS-1-Functions-01dd8400b85f40d083966908acbfa184"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  LESSON
                </a>
                <a
                  class="btn border-right rounded-0 px-4 py-3"
                  href="/curriculum/2"
                >
                  CHALLENGES
                </a>
              </div>
            </div>
          </div>
          <div
            class="alerts-container "
          >
            <div
              class="alert d-flex justify-content-between mt-3 bg-primary text-white"
              role="alert"
            >
              <div>
                <img
                  class="mr-3 alert-icon"
                  src="/assets/curriculum/icons/icon-tip.svg"
                />
                Set up your computer to submit challenges. 
                <a
                  class="text-white"
                  href="https://www.notion.so/JS-0-Foundations-a43ca620e54945b2b620bcda5f3cf672#b45ed85a95e24c9d9fb784afb7a46bcc"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  View Instructions
                </a>
              </div>
              <button
                class="alert-dismiss"
                data-testid="dismiss-info"
                role="dismiss"
              >
                <img
                  src="/assets/curriculum/icons/dismiss-info.svg"
                />
              </button>
            </div>
            <div
              class="alert d-flex justify-content-between mt-3 alert-danger"
              role="alert"
            >
              <div>
                <img
                  class="mr-3 alert-icon"
                  src="/assets/curriculum/icons/exclamation.svg"
                />
                Please upgrade your CLI client by running npm update c0d3. 
                
              </div>
              <button
                class="alert-dismiss"
                data-testid="dismiss-urgent"
                role="dismiss"
              >
                <img
                  src="/assets/curriculum/icons/dismiss-urgent.svg"
                />
              </button>
            </div>
          </div>
          <div
            class="row challenge-display mt-3"
          >
            <div
              class="col-4"
            >
              <div
                class="card mb-2 challenge-title-card challenge-title-card--done"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    1. Incremental Closure
                  </div>
                  <img
                    height="25px"
                    src="/assets/curriculum/icons/checkmark.svg"
                    width="25px"
                  />
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card challenge-title-card--done"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    2. Closure x times
                  </div>
                  <img
                    height="25px"
                    src="/assets/curriculum/icons/checkmark.svg"
                    width="25px"
                  />
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card challenge-title-card--active shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    3. Repeated String
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    4. Calling input function
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    5. Call function X Times
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    6. isPrime
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    7. Letter Find
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    8. Letter Count
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    9. Letter Map
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    10. Letter Loop
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    11. Delayed Closure
                  </div>
                </div>
              </div>
              <div
                class="card mb-2 challenge-title-card shadow-sm border-0"
                data-testid="challenge-title"
              >
                <div
                  class="card-body d-flex justify-content-between"
                >
                  <div>
                    12. Delay 1 and 2
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col-8"
            >
              <div
                class="card shadow-sm border-0"
              >
                <div
                  class="card-body challenge-question"
                >
                  <h1
                    class="card-title"
                    data-testid="challenge-question-title"
                  >
                    Repeated String
                  </h1>
                  <div
                    class="bg-light p-3 mt-3"
                    data-testid="challenge-question-description"
                  >
                    <div>
                      <p>
                        Write a function called solution that takes in 2 parameters, a number and string and returns the string repeated input parameter number of times
                      </p>
                      <pre>
                        <code>
                          result = solution(3, "abc") // "abcabcabc"
result = solution(0, "hello") // ""
                        </code>
                      </pre>
                      <p>
                        Reminder for students with prior experience: you are not allowed to use 
                        <code>
                          for
                        </code>
                         and 
                        <code>
                          while
                        </code>
                         loops. 
                        <a
                          href="https://www.notion.so/garagescript/Recursion-973993e8ed144d8a87cb5695fd1ea8b1"
                        >
                          Reason
                        </a>
                      </p>
                    </div>
                  </div>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="mt-5 font-weight-light text-center pb-5 text-muted"
    >
      © Copyright 2020 GarageScript
      <br />
      All Rights Reserved
    </footer>
  </div>
</div>
`;

exports[`Lesson Page Should return Internal server Error if alerts or lessons are missing 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
              rel="noopener noreferrer"
              target="_blank"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
            <a
              class="nav-item nav-link"
              href="/contributors"
            >
              Contributors
            </a>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-none"
                type="button"
              >
                Admin
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex"
      >
        <a
          class="btn btn-secondary border overflow-hidden p-2 text-truncate"
          href="/profile/fakeusername"
        >
           fakeusername
        </a>
        <button
          class="btn border ml-2"
          style="color: rgb(41, 41, 41);"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="d-flex col-sm-3 align-items-center justify-content-center"
        >
          <div
            class="text-center mt-3"
          >
            <div
              style="color: rgb(41, 41, 41); font-size: 1.5rem;"
            >
              <strong>
                Internal server error
              </strong>
            </div>
            <div
              style="color: rgb(41, 41, 41); font-size: 1rem;"
            >
              Bad data
            </div>
            <a
              class="btn btn-primary py-3 px-5 mt-3"
              href="/"
            >
              <h3
                class="font-weight-bold"
              >
                Back
              </h3>
            </a>
          </div>
        </div>
        <div
          class="mb-3 col-sm-9"
        >
          <div
            style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="display: block; box-sizing: border-box; padding-top: 100%;"
            />
            <img
              decoding="async"
              sizes="100vw"
              src="/_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=3840&q=75"
              srcset="/_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=640&q=75 640w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=750&q=75 750w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=828&q=75 828w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=1080&q=75 1080w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=1200&q=75 1200w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=1920&q=75 1920w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=2048&q=75 2048w, /_next/image?url=%2Fassets%2Ferrors%2F500.svg&w=3840&q=75 3840w"
              style="visibility: inherit; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;
